### DISCUSSION TITLE
Open popup as new instance of ChromiumWebBrowser in WinForms

### DISCUSSION THREAD

**Discussion by GITMoor** [cefsharp/CefSharp#4308] [Q&A]
Is there a way to implement something like [**this**](https://github.com/cefsharp/CefSharp/pull/3529#issuecomment-828934239) but instead of ChromiumHostControl return a new instance of a ChromiumWebBrowser? The problem is that when you open a pop in a new tab of [**EasyTabs**](https://github.com/lstratman/EasyTabs) the popup is conected to ChromiumWebBrowser from witch it was opened, and if you torn tab into a new window the popups opened from that popup will appear in the torned window where  ChromiumWebBrowser is located. Here is a vid of problem

https://user-images.githubusercontent.com/107041444/202201363-beb2d14a-ba6d-463c-9ba2-eb0d94ea5293.mp4

**Comment by amaitland** [comment]
> The problem is that when you open a pop in a new tab of [EasyTabs](https://github.com/lstratman/EasyTabs) the popup is conected to ChromiumWebBrowser from witch it was opened, and if you torn tab into a new window the popups opened from that popup will appear in the torned window where ChromiumWebBrowser is located.

How would hosting the popup in a ChromiumWebBrowser instance change this behaviour exactly? Sounds like an implementation detail of how you've implemented the tabs.

**Reply by GITMoor** [reply]
Since then i did some investigation, and now the problem is solved. But i have one more question. When closing window with  ChromiumWebBrowser to prevent ChromiumHostControl's stop working im setting its parent to null, but how can i know when all popups being closed to dispose ChromiumWebBrowser later?

**Reply by amaitland** [reply]
> but how can i know when all popups being closed to dispose ChromiumWebBrowser later?

You'll need to keep a count of open popups, increment count in OnPopupCreated, decrement count in OnPopupDestroyed.

Nothing publicly accessible at the moment. There's a dictionary internally, it's deep within the `C++` side, something we could potentially expose in a controlled way.

https://github.com/cefsharp/CefSharp/blob/cefsharp/107/CefSharp.Core.Runtime/Internals/ClientAdapter.h#L47

PR's welcome.

**Reply by amaitland** [reply]
Alternatively [LifeSpanHandler](https://github.com/cefsharp/CefSharp/blob/cefsharp/107/CefSharp.WinForms/Handler/LifeSpanHandler.cs#L72) could be improved to keep count and provide the number of active popups in `OnPopupDestroyed` or something similar.
